torch/testing/_internal/custom_op_db.py

```diff
@@ -0,0 +1,112 @@
+import torch
+from torch.testing._internal.opinfo.core import (
+    OpInfo,
+)
+from torch.testing._internal.common_dtype import all_types_and
+import numpy as np
+from torch._custom_op import CustomOp
+from torch.testing._internal.autograd_function_db import (
+    sample_inputs_numpy_cube,
+    sample_inputs_numpy_mul,
+    sample_inputs_numpy_sort,
+    sample_inputs_numpy_take,
+)
+
+# Note: [custom op db]
+#
+# This is a collection of custom operator test cases written as OpInfos
+# so they can easily be consumed by OpInfo-based tests to check if subsystems
+# support them correctly.
+
+def to_numpy(tensor):
+    return tensor.cpu().numpy()
+
+numpy_cube = CustomOp.define('_torch_testing::numpy_cube(Tensor x) -> (Tensor, Tensor)')
```

Comment:
OK, imma bikeshed this more.

The downside to implementing the object this way is `numpy_cube` will not have the correct `__name__` and `__module__` stored on it, because obviously CustomOp.define has no idea where its result is going to be bound. The way we handled this in primtorch is to define a stub function and then decorate it, e.g.,

```
@_make_elementwise_unary_reference(ELEMENTWISE_TYPE_PROMOTION_KIND.COMPLEX_TO_FLOAT)
def abs(a):
    return prims.abs(a)
```

instead of

```
abs = _make_elementwise_unary_reference(ELEMENTWISE_TYPE_PROMOTION_KIND.COMPLEX_TO_FLOAT, prims.abs)
```

This is troublesome for your use case, because your custom ops don't really have a mandatory function body. Or do they? Forcing a meta implementation doesn't seem like the worst thing in the world.

Reply:
An extra benefit to doing it this way, is the function you feed into the decorator can provide the namespace (assuming we have some suitable mangling scheme from Python module names to TS namespaces, which I think we should!) and the function name, so you don't have to repeat yourself here.

```
@CustomOp.define_meta('(Tensor x) -> (Tensor, Tensor)')
def numpy_cube(x):
    return x.clone(), x.clone()
```

And if you want to get really fancy, we could derive the schema from mypy type annotations:

```
@CustomOp.define_meta()
def numpy_cube(x: Tensor) -> Tuple[Tensor, Tensor]:
    return x.clone(), x.clone()
```

Reply:
Let me think about this some more. Some initial responses:

> because your custom ops don't really have a mandatory function body. Or do they? Forcing a meta implementation doesn't seem like the worst thing in the world.

If we go this route, I'm leaning towards the user defining a dummy function instead of us having a default:
```py
@CustomOp.define('(Tensor x) -> (Tensor, Tensor)')
def numpy_cube(x):
    pass
```
or take the route of a default. Forcing everyone to have a meta implementation seems good for PT2, but has the downside of that it's not always possible to write the meta implementation (data-dependent rules need to go through a FakeTensor registration, unless we combine the FakeTensor and meta registrations).

> (assuming we have some suitable mangling scheme from Python module names to TS namespaces, which I think we should!) and the function name, so you don't have to repeat yourself here.

I've thought about the mangling the python module name to TS namespace. The downside is that the user (if they're working with the FX graphs or need to use the raw torch.library API) needs to know about the mangling scheme. But maybe this isn't too bad if we document what is going on

Reply:
OK, let's do placeholder. I suggest the idiomatic style should be ellipses rather than pass.

> I've thought about the mangling the python module name to TS namespace. The downside is that the user (if they're working with the FX graphs or need to use the raw torch.library API) needs to know about the mangling scheme. But maybe this isn't too bad if we document what is going on. The pros are that the user really doesn't need to worry about namespace collision and we take care of namespacing for them

Yeah, let's define a permanent mangling scheme (it probably can be pretty simple; just need a prefix like py, an encoding scheme for dot, and an escape for literal occurrence of dot.)

Reply:
> Yeah, let's define a permanent mangling scheme (it probably can be pretty simple; just need a prefix like py, an encoding scheme for dot, and an escape for literal occurrence of dot.)

@ezyang our options seem like either we do the mangling, or we teach schema about nested namespaces. I'm not sure how difficult the latter is right now, but there are some other things we need to do if we go the mangling approach:
- I want to put some hook into torch_dispatch so that when torch.ops.PYmylibraryDOTcustom_ops_V0.foo.default gets returned to Python, we actually get back mylibrary.custom_ops.foo instead of the OpOverload and also have the demangled form show up in the FX graph (so that it's easier to manipulate). Are we OK with this? 
- People who are trying to override the operation from C++ or using torch.library need to work with the mangled name ("PYmylibraryDOTcustom_ops_V0::foo"), unless we want to let them somehow use the unmangled name.

Also, if we are inferring the name of the operator from the python function, then, we need to figure out some initial plan to deal overloads (at least the out= variant) for now, since that will be the first thing that people ask us for.

Reply:
Latter puts you into getting TorchScript on board, which just seems like a lot of pain.

> I want to put some hook into torch_dispatch so that when torch.ops.PYmylibraryDOTcustom_ops_V0.foo.default gets returned to Python, we actually get back mylibrary.custom_ops.foo instead of the OpOverload and also have the demangled form show up in the FX graph (so that it's easier to manipulate). Are we OK with this?

Yes, this sounds good. Since you're eschewing OpOverload anyway you probably wanted to do this anyway, so that you get your new object type. But whatever API you expose needs to be compatible with OpOverload...

> People who are trying to override the operation from C++ or using torch.library need to work with the mangled name ("PYmylibraryDOTcustom_ops_V0::foo"), unless we want to let them somehow use the unmangled name.

This seems fine

> Also, if we are inferring the name of the operator from the python function, then, we need to figure out some initial plan to deal overloads (at least the out= variant) for now, since that will be the first thing that people ask us for.

Seems fine, we can have a convention for out= variant overload name.